api/test/e2e/route_test.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package e2e
+
+import (
+	"net/http"
+	"testing"
+	"time"
+)
+
+func TestRoute_Host(t *testing.T) {
```

Comment:
The current test cases are pipelined and are not divided into functions, which makes it difficult to understand and modify. 
I'm not sure that the test cases of go are written like this? In my python and Lua world, it is a function to complete an independent test.

Reply:
Coder can run their tests parallel just using the `testing.T.Parallel`.

Reply:
Coder can run their tests parallel just using the `testing.T.Parallel`.

Reply:
Cases in this test can be more concise, specifically, all the shape of cases are same, only the contents are different, so why not write them just in a loop and defines these case-related artifacts separately, which makes the code base clean and more readable, whats' more, it's easy to run cases parallelly.

```go
func TestRoute_Host(t *testing.T) {
    cases := ...

    for _, case := range cases {
        t.Run("route", func(t *testing.T) {
            t.Parallel()
            // test case blah blah blah
        })
    }
}
```

Reply:
If you agree with this comment, kindly mark this PR as approval, or keep requesting changes.

![image](https://user-images.githubusercontent.com/2106987/97869265-23c89500-1d4c-11eb-8fb0-2e99e198d12e.png)